File: src/main/java/io/github/takenoha/towerdefense/persistence/ResourcePickupFeedback.java
package io.github.takenoha.towerdefense.persistence;

import java.util.Objects;
import java.util.UUID;

/** Values used by the main-thread pickup feedback after a committed claim. */
public record ResourcePickupFeedback(
        UUID eventId,
        UUID playerId,
        ResourceType resourceType,
        int claimedQuantity,
        long eventPlayerTotal,
        long teamBalance) {
    public ResourcePickupFeedback {
        Objects.requireNonNull(eventId, "eventId");
        Objects.requireNonNull(playerId, "playerId");
        Objects.requireNonNull(resourceType, "resourceType");
        if (claimedQuantity <= 0 || eventPlayerTotal < 0 || teamBalance < 0) {
            throw new IllegalArgumentException("pickup feedback quantities are invalid");
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/persistence/ResourcePickupFeedback.kt
package io.github.takenoha.towerdefense.persistence

import java.util.Objects
import java.util.UUID
import kotlin.jvm.JvmRecord

/** Values used by the main-thread pickup feedback after a committed claim. */
@JvmRecord
data class ResourcePickupFeedback(
    val eventId: UUID,
    val playerId: UUID,
    val resourceType: ResourceType,
    val claimedQuantity: Int,
    val eventPlayerTotal: Long,
    val teamBalance: Long,
) {
    init {
        Objects.requireNonNull(eventId, "eventId")
        Objects.requireNonNull(playerId, "playerId")
        Objects.requireNonNull(resourceType, "resourceType")
        if (claimedQuantity <= 0 || eventPlayerTotal < 0L || teamBalance < 0L) {
            throw IllegalArgumentException("pickup feedback quantities are invalid")
        }
    }
}
